File: backend/app/routers/bench.py
"""The bench surface — thin HTTP over app.bench (datasets → gold → run → report).

Long operations (gold drafting, the run itself) stream NDJSON, same protocol as
the graph page's upload/dream streams, so the Bench page narrates progress live.
"""

from __future__ import annotations

import json
import re

from fastapi import APIRouter, HTTPException
from fastapi.responses import PlainTextResponse, StreamingResponse
from pydantic import BaseModel

import shutil

from app import bench
from app.bench import estimate as bench_estimate
from app.bench import fetch, goldgen, jobs, runner, store

router = APIRouter(prefix="/bench")


def _line(obj: dict) -> bytes:
    return (json.dumps(obj) + "\n").encode("utf-8")


@router.get("/datasets")
def datasets() -> dict:
    return {"datasets": bench.list_datasets(), "raw_dir": str(store.RAW_DIR)}


class DownloadBody(BaseModel):
    url: str


@router.post("/download")
def download(body: DownloadBody) -> StreamingResponse:
    """Fetch a dataset file from a URL into the raw dir, streaming progress."""
    def stream():
        for ev in fetch.download(body.url):
            yield _line(ev)

    return StreamingResponse(stream(), media_type="application/x-ndjson")


class DeleteDatasetBody(BaseModel):
    name: str


@router.post("/delete-dataset")
def delete_dataset(body: DeleteDatasetBody) -> dict:
    """Remove a dataset's raw file(s) and derived work dir (corpus, gold, manifest,
    runs). Graph saves built from it are NOT touched — manage those on the graph page."""
    name = body.name.strip()
    if not re.fullmatch(r"[A-Za-z0-9._-]+", name):
        raise HTTPException(status_code=400, detail="Bad dataset name.")
    removed = []
    for suffix in (".jsonl", ".json"):
        f = store.RAW_DIR / f"{name}{suffix}"
        if f.exists():
            f.unlink()
            removed.append(f.name)
    work = store.WORK_DIR / name
    if work.exists():
        shutil.rmtree(work)
        removed.append(f"{name}/ (prepared corpus, gold, runs)")
    if not removed:
        raise HTTPException(status_code=404, detail="No such dataset.")
    return {"removed": removed}


class PrepareBody(BaseModel):
    dataset: str
    cap_tokens: int = 100_000


@router.post("/prepare")
def prepare(body: PrepareBody) -> dict:
    try:
        prepared = bench.prepare(body.dataset, max(10_000, body.cap_tokens))
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc))
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    return {"prepared": prepared, **bench.dataset_status(body.dataset)}


class GoldGenBody(BaseModel):
    dataset: str
    total: int = 24


@router.post("/goldgen")
def goldgen_stream(body: GoldGenBody) -> StreamingResponse:
    async def stream():
        async for ev in goldgen.generate(body.dataset, max(1, min(body.total, 100))):
            yield _line(ev)

    return StreamingResponse(stream(), media_type="application/x-ndjson")


class VerifyBody(BaseModel):
    dataset: str


@router.post("/goldverify")
def goldverify_stream(body: VerifyBody) -> StreamingResponse:
    async def stream():
        async for ev in goldgen.verify(body.dataset):
            yield _line(ev)

    return StreamingResponse(stream(), media_type="application/x-ndjson")


@router.get("/gold")
def get_gold(dataset: str) -> dict:
    return {"questions": store.load_gold(dataset)}


class GoldBody(BaseModel):
    dataset: str
    questions: list[dict]


@router.put("/gold")
def put_gold(body: GoldBody) -> dict:
    keep = []
    for q in body.questions:
        if not (q.get("question") and q.get("answer")):
            continue
        q["status"] = q.get("status") if q.get("status") in ("draft", "approved") else "draft"
        keep.append(q)
    store.save_gold(body.dataset, keep)
    return {"saved": len(keep),
            "approved": sum(1 for q in keep if q["status"] == "approved")}


class RunBody(BaseModel):
    dataset: str
    configs: list[str] = list(runner.CONFIGS)
    extract_model: str | None = None  # graph (and LightRAG) builder
    answer_model: str | None = None   # the generator every config shares
    judge_model: str | None = None    # verdict model (on the judge endpoint if set)
    context_model: str | None = None  # chunk contextualizer (RAG build blurbs)
    scope: str = "auto"  # "auto" = follow the dataset (scope to each question's doc_id when it has one); "corpus" = force whole-corpus


@router.post("/run")
async def run(body: RunBody) -> StreamingResponse:
    """Start a run as a DETACHED job and tail it. Closing the tab only drops the
    tail — the run keeps going server-side; GET /bench/job reattaches to it."""
    job = jobs.start(body.dataset, "run", runner.run_bench(
        body.dataset, body.configs,
        extract_model=body.extract_model, answer_model=body.answer_model,
        judge_model=body.judge_model, context_model=body.context_model,
        scope=body.scope,
    ))

    async def stream():
        if job is None:
            yield _line({"phase": "error",
                         "detail": "A run is already in progress for this dataset — reattach "
                                   "to it (reload the page) or pause it first. Starting a "
                                   "second run would rebuild the same memory and pay twice."})
            return
        async for ev in job.tail():
            yield _line(ev)

    return StreamingResponse(stream(), media_type="application/x-ndjson")


@router.get("/job")
async def job_for_dataset(dataset: str) -> dict:
    """The dataset's newest job (running or finished) so a reopened page can reattach."""
    job = jobs.active_for(dataset) or jobs.latest_for(dataset)
    return {"job": job.info() if job else None}


@router.get("/job/{job_id}/stream")
async def job_stream(job_id: str, since: int = 0) -> StreamingResponse:
    job = jobs.get(job_id)
    if job is None:
        raise HTTPException(status_code=404, detail="No such job (server restarted?) — "
                                                    "press Run to resume; nothing is re-paid.")

    async def stream():
        async for ev in job.tail(since):
            yield _line(ev)

    return StreamingResponse(stream(), media_type="application/x-ndjson")


@router.post("/job/{job_id}/stop")
async def job_stop(job_id: str) -> dict:
    job = jobs.get(job_id)
    if job is None:
        raise HTTPException(status_code=404, detail="No such job.")
    job.stop()
    return {"stopped": True}


@router.get("/jobs")
def all_jobs() -> dict:
    """Every running job across datasets — lets the page show the whole overnight
    campaign, not just the selected dataset's job."""
    return {"jobs": [j.info() for j in jobs.all_active()]}


@router.get("/estimate")
def estimate(dataset: str, configs: str = "", extract_model: str = "",
             context_model: str = "", answer_model: str = "") -> dict:
    """Cost ballpark for a run BEFORE it starts (±2× until calibrated). `configs`
    is comma-separated; empty = every config. Model overrides mirror the run's
    pickers so the gate prices what will actually launch."""
    wanted = [c for c in configs.split(",") if c] or list(runner.CONFIGS)
    return bench_estimate.estimate(dataset, wanted, extract_model or None,
                                   context_model or None, answer_model or None)


class RejudgeBody(BaseModel):
    dataset: str
    run_id: str
    judge_model: str | None = None


@router.post("/rejudge")
async def rejudge(body: RejudgeBody) -> StreamingResponse:
    """Re-score a stored run with the current judge/gold — no generation re-paid.
    Also a detached job, so a closed tab doesn't abandon paid verdicts mid-pass."""
    job = jobs.start(body.dataset, "rejudge",
                     runner.rejudge_run(body.dataset, body.run_id, body.judge_model))

    async def stream():
        if job is None:
            yield _line({"phase": "error",
                         "detail": "A job is already running for this dataset — wait for it "
                                   "or pause it first."})
            return
        async for ev in job.tail():
            yield _line(ev)

    return StreamingResponse(stream(), media_type="application/x-ndjson")


@router.get("/runs")
def runs(dataset: str) -> dict:
    return {"runs": store.list_runs(dataset)}


class DeleteRunBody(BaseModel):
    dataset: str
    run_id: str


@router.post("/delete-run")
def delete_run(body: DeleteRunBody) -> dict:
    """Remove one run's report from the dataset's workdir. The local archive copy
    (bench_archive/) is kept — deleting from the UI never destroys the last record."""
    if not re.fullmatch(r"[A-Za-z0-9._-]+", body.dataset):
        raise HTTPException(status_code=400, detail="Bad dataset name.")
    if not re.fullmatch(r"[A-Za-z0-9._-]+", body.run_id):
        raise HTTPException(status_code=400, detail="Bad run id.")
    if not store.delete_run(body.dataset, body.run_id):
        raise HTTPException(status_code=404, detail="No such run.")
    return {"deleted": body.run_id}


@router.get("/report")
def report(dataset: str, run_id: str, full: bool = False) -> dict:
    """The report; `full=true` includes the raw per-question records (large — at
    1000 questions they are megabytes, so the page never loads them by default)."""
    r = store.load_run(dataset, run_id)
    if not r:
        raise HTTPException(status_code=404, detail="No such run.")
    return r if full else {**r, "records": []}


@router.get("/report.md", response_class=PlainTextResponse)
def report_md(dataset: str, run_id: str) -> str:
    md = store.load_run_markdown(dataset, run_id)
    if md is None:
        raise HTTPException(status_code=404, detail="No such run.")
    return md

File: frontend/src/api/bench.js
// The bench surface: datasets → prepare → gold questions → run (streamed) → report.
import { API_BASE, asJson, authFetch, readNdjsonStream } from "./client.js";

export const listDatasets = () => authFetch(`${API_BASE}/bench/datasets`).then(asJson);

export const prepareDataset = (dataset, capTokens) =>
  authFetch(`${API_BASE}/bench/prepare`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ dataset, cap_tokens: capTokens }),
  }).then(asJson);

export const getGold = (dataset) =>
  authFetch(`${API_BASE}/bench/gold?dataset=${encodeURIComponent(dataset)}`).then(asJson);

export const putGold = (dataset, questions) =>
  authFetch(`${API_BASE}/bench/gold`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ dataset, questions }),
  }).then(asJson);

// Streams {phase:"start"|"progress"|"done"|"error", ...} while the LLM drafts questions.
export async function goldgenStream(dataset, total, onEvent) {
  const res = await authFetch(`${API_BASE}/bench/goldgen`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ dataset, total }),
  });
  await readNdjsonStream(res, onEvent);
}

// Streams the auto-verify pass over draft questions: mechanical evidence check +
// one judge call each; passes get approved, failures stay draft with a `flag`.
export async function goldverifyStream(dataset, onEvent) {
  const res = await authFetch(`${API_BASE}/bench/goldverify`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ dataset }),
  });
  await readNdjsonStream(res, onEvent);
}

// Streams the whole run: build (or build_skip) → answers → judging → report.
// The run is a DETACHED server job — `signal` only drops this tail (the run keeps
// going; getActiveJob + attachJobStream reattach to it). Use stopJob to actually pause.
export async function runStream(
  { dataset, configs, extractModel, answerModel, judgeModel, contextModel, scope, signal },
  onEvent,
) {
  const res = await authFetch(`${API_BASE}/bench/run`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      dataset,
      configs,
      extract_model: extractModel || null,
      answer_model: answerModel || null,
      judge_model: judgeModel || null,
      context_model: contextModel || null,
      scope: scope || "auto",
    }),
    signal,
  });
  await readNdjsonStream(res, onEvent);
}

// Streams a dataset download from a URL into the raw dir:
// {phase:"download_start"|"progress"|"done"|"error", ...}
export async function downloadStream(url, onEvent) {
  const res = await authFetch(`${API_BASE}/bench/download`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ url }),
  });
  await readNdjsonStream(res, onEvent);
}

// Re-scores a stored run's answers with the current judge + gold (no generation
// cost). `judgeModel` overrides the verdict model for this pass.
export async function rejudgeStream(dataset, runId, judgeModel, onEvent) {
  const res = await authFetch(`${API_BASE}/bench/rejudge`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ dataset, run_id: runId, judge_model: judgeModel || null }),
  });
  await readNdjsonStream(res, onEvent);
}

// The dataset's newest job (running or finished) — lets a reopened page reattach
// to a run that survived the tab.
export const getActiveJob = (dataset) =>
  authFetch(`${API_BASE}/bench/job?dataset=${encodeURIComponent(dataset)}`).then(asJson);

// Every running job across all datasets — the overnight-campaign view.
export const listAllJobs = () =>
  authFetch(`${API_BASE}/bench/jobs`).then(asJson);

// Tail a job's event log from `since` (0 = full replay), following live until done.
export async function attachJobStream(jobId, since, onEvent, signal) {
  const res = await authFetch(
    `${API_BASE}/bench/job/${encodeURIComponent(jobId)}/stream?since=${since || 0}`,
    { signal },
  );
  await readNdjsonStream(res, onEvent);
}

// Actually pause the server-side job (aborting the tail alone never stops the work).
export const stopJob = (jobId) =>
  authFetch(`${API_BASE}/bench/job/${encodeURIComponent(jobId)}/stop`, { method: "POST" }).then(asJson);

export const deleteDataset = (name) =>
  authFetch(`${API_BASE}/bench/delete-dataset`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name }),
  }).then(asJson);

export const killRun = (dataset) =>
  authFetch(`${API_BASE}/bench/kill`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ dataset }),
  }).then(asJson);

export const deleteRun = (dataset, runId) =>
  authFetch(`${API_BASE}/bench/delete-run`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ dataset, run_id: runId }),
  }).then(asJson);

export const getEstimate = (dataset, configs, models = {}) =>
  authFetch(
    `${API_BASE}/bench/estimate?dataset=${encodeURIComponent(dataset)}&configs=${configs.join(",")}` +
      `&extract_model=${encodeURIComponent(models.extract || "")}` +
      `&context_model=${encodeURIComponent(models.context || "")}` +
      `&answer_model=${encodeURIComponent(models.answer || "")}`,
  ).then(asJson);

export const listRuns = (dataset) =>
  authFetch(`${API_BASE}/bench/runs?dataset=${encodeURIComponent(dataset)}`).then(asJson);

export const getReport = (dataset, runId) =>
  authFetch(
    `${API_BASE}/bench/report?dataset=${encodeURIComponent(dataset)}&run_id=${encodeURIComponent(runId)}`,
  ).then(asJson);

// Fetches report.md through the authenticated API and saves it as a file. A bare
// <a href> link cannot carry the session header, so it bounces off the auth gate
// ("Not signed in") — the download has to go through authFetch and a blob.
export async function downloadReportMd(dataset, runId) {
  const res = await authFetch(
    `${API_BASE}/bench/report.md?dataset=${encodeURIComponent(dataset)}&run_id=${encodeURIComponent(runId)}`,
  );
  if (!res.ok) {
    let detail = `HTTP ${res.status}`;
    try {
      detail = (await res.json()).detail ?? detail;
    } catch {
      /* keep generic */
    }
    throw new Error(detail);
  }
  const url = URL.createObjectURL(await res.blob());
  const a = document.createElement("a");
  a.href = url;
  a.download = `${dataset}-${runId}-report.md`;
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(url);
}
